### ISSUE TITLE
useChat maxRoundTrips breaks in 3.3.21

### ISSUE THREAD

**Issue by stephenkalnoske** [vercel/ai#2856]
### Description

maxRoundTrips for toolcalls in the useChat hook on client side breaks in 3.3.21. It will not re-call the chat endpoint. I understand that there's maxRoundTrips in the streamText code now, but shouldn't it be backwards-compatible?

Downgrading to 3.3.20 works fine.

**Comment by lesenelir** [comment]
I have the same problem.

**Comment by danielzohar** [comment]
It may be the same problem or a related one -

When the following conditions are met:
- `useChat` is used on the client
- `streamText` is used on the server 
- `maxToolRoundtrips` on the server is set to a non zero value
-  A tool that does a blocking async execution (e.g. `await fetch...`) is being called on the server

The message stream on the client appears only until the tool call and the following error is logged:
```
Error: tool_result must be preceded by a tool_call
    at processDataProtocolResponse (index.mjs:855:15)
    at async callChatApi (index.mjs:1023:14)
    at async getStreamedResponse (index.mjs:222:10)
    at async processChatStream (index.mjs:1235:42)
    at async eval (index.mjs:337:9)
```    

This problem does NOT occur when the tool call returns immediately like [in the example docs](https://github.com/vercel/ai/blob/main/content/docs/03-ai-sdk-core/15-tools-and-tool-calling.mdx#example-streamtext).

## Code example
I created a proof of concept repository to demonstrate the issue here https://github.com/danielzohar/usechat-streamtext-tools-issue-poc

**Comment by lesenelir** [comment]
> It may be the same problem or a related one -
> 
> When the following conditions are met:
> 
> * `useChat` is used on the client
> * `streamText` is used on the server
> * `maxToolRoundtrips` on the server is set to a non zero value
> * A tool that does a blocking async execution (e.g. `await fetch...`) is being called on the server
> 
> The message stream on the client appears only until the tool call and the following error is logged:
> 
> ```
> Error: tool_result must be preceded by a tool_call
>     at processDataProtocolResponse (index.mjs:855:15)
>     at async callChatApi (index.mjs:1023:14)
>     at async getStreamedResponse (index.mjs:222:10)
>     at async processChatStream (index.mjs:1235:42)
>     at async eval (index.mjs:337:9)
> ```
> 
> This problem does NOT occur when the tool call returns immediately like [in the example docs](https://github.com/vercel/ai/blob/main/content/docs/03-ai-sdk-core/15-tools-and-tool-calling.mdx?rgh-link-date=2024-09-01T10%3A46%3A43Z#example-streamtext).
> 
> ## Code example
> I created a proof of concept repository to demonstrate the issue here [danielzohar/usechat-streamtext-tools-issue-poc](https://github.com/danielzohar/usechat-streamtext-tools-issue-poc)


If `maxToolRoundtrips` is not zero in useChat, but `maxToolRoundtrips` is zero in streamText, there will be the same problem.

```
Error: tool_result must be preceded by a tool_call
    at processDataProtocolResponse (process-data-protocol-response.ts:241:15)
    at async callChatApi (call-chat-api.ts:114:14)
    at async getStreamedResponse (use-chat.ts:139:10)
    at async processChatStream (process-chat-stream.ts:33:42)
    at async use-chat.ts:341:9
```

It seems that the execute function in the tool cannot perform asynchronous tasks.

**Comment by lgrammel** [comment]
Thanks @danielzohar for the investigation. I've reproduced the bug, the `roundtrip-finish` event is sent too early when there are ongoing tool calls: https://github.com/vercel/ai/pull/2859/files

**Comment by lgrammel** [comment]
Fixed in [ai@3.3.22](https://github.com/vercel/ai/releases/tag/ai%403.3.22)

**Comment by sh0umik** [comment]
@lgrammel working `maxToolRoundtrips` is working now on `useChat` i was getting this type of error while follong the RAG chatbot example

```
MessageConversionError [AI_MessageConversionError]: ToolInvocation must have a result: {"state":"call","toolCallId":"call_YXBWSPk9Mek2TsEQBnQIRVx6","toolName":"getInformation","args":{"question":"What is the user's favorite fruit?"}}
```

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This is a confirmed regression in existing useChat tool-roundtrip behavior, reproduced and fixed by a maintainer.

### Evidence
- The reporter identifies a regression in ai@3.3.21: useChat no longer recalls the chat endpoint for tool calls, while downgrading to 3.3.20 restores the behavior.
- Comments provide reproducible errors involving asynchronous tool execution, including “tool_result must be preceded by a tool_call.”
- A maintainer reproduced the issue and identified that the `roundtrip-finish` event was sent too early while tool calls were still ongoing.
- The maintainer confirmed the fix was released in ai@3.3.22.
